Springboard -- Data Science Career Program -- Capstone Project #1: Real Estate Price Predictor

Data Wrangling -- By Kevin Cole -- March 2020

This document describes data wrangling steps and the code to support it. Capstone proposal can be found below.

Initial Proposal
https://github.com/ABitNutty/Capstone-1/blob/master/Capstone%201%20Proposal.pdf


<b>The Data</b>

The data set is a premade .CSV file found on the Kaggle website. 
    https://www.kaggle.com/c/house-prices-advanced-regression-techniques/data

The data is contained within the download package as a file called ‘train.csv’. The dataset contains 1460 data points, each being the records for a property sold in Ames, Iowa. There are 79 variables available for analysis, along with a column for the target variable of sale price. 

In [1]:
#Import packages to be used
import pandas as pd
import numpy as np

# Imports raw data
data = pd.read_csv('train.csv')
data = data.set_index('Id')

<b>Initial variable typing</b>

This dataset was examined and each variable was placed into one of three categories; Categorical variables based on strings, categorical variables that are coded numerically, or noncategorical (numeric) based on what the data was describing. These descions were made based on visual inspections of the data as well as the documention contianed in the data download package (data_description.txt)

After determining the type of each variable that information was saved off in an initialization file for later access. Here, we will import that .CSV file to continue building the dataframe that will help understand the structure of the data. 

In [2]:
# Imports initialization .CSV for Variable exploration and recoring
data_information = pd.read_csv('Variable_Information.csv', index_col='Column Name')
del data_information['Unnamed: 0']
data_information.index.name = 'Column Name'

# Creates dataframes for each of the variable types
numeric = data_information[data_information['Type'] == 'Numeric']
categorical_string = data_information[data_information['Type'] == 'Categorical String']
categorical_numeric = data_information[data_information['Type'] == 'Categorical Numeric']

# Initialize dataframe for future inputs
data_information['Category Distribution'] = pd.Series()
data_information['Nan_count'] = pd.Series()
data_information['Nan_percent'] = pd.Series()
data_information['Summary Statistics'] = pd.Series()

#Prints data_information to understand what we are building
data_information.head()

Type  Category Distribution  Nan_count  \
Column Name                                                          
MSSubClass   Categorical Numeric                    NaN        NaN   
MSZoning      Categorical String                    NaN        NaN   
LotFrontage              Numeric                    NaN        NaN   
LotArea                  Numeric                    NaN        NaN   
Street        Categorical String                    NaN        NaN   

             Nan_percent  Summary Statistics  
Column Name                                   
MSSubClass           NaN                 NaN  
MSZoning             NaN                 NaN  
LotFrontage          NaN                 NaN  
LotArea              NaN                 NaN  
Street               NaN                 NaN

We can see that each variable in the dataset, a row is created in data_information. The type of that variable was imported from the initialization file. The rest of the categories will be populated below. 

<b> Categorical Analysis </b>

For categorical variables, there are a few things that we want to know. By creating a distribution of occurrences for each category, we not only understand how many distinct categories exist but can reasonably make frequency histograms from the results. 

By specifiying dropna=False we include the NaN frequencies in the distribution for counting as well. This information is also useful in determing the percentage of missing data. 

In [3]:
# Create number of distict events for each category in a categorical variable (Categorical String)
for variable_name, notused in categorical_string.iterrows():
    # Frequency Distribution
    distribution = data[variable_name].value_counts(dropna= False).to_dict()
    
    if np.nan in list(distribution.keys()):
        #Count NaN elements and add to distribution
        data_information.loc[variable_name, 'Nan_count'] = distribution[np.nan]
    else:
        data_information.loc[variable_name, 'Nan_count'] = 0
    
    # This line of code works as intended but throws SettingWithCopyWarning. Better way to store this?
    data_information['Category Distribution'][variable_name] = (list(distribution.keys()), list(distribution.values()))

    
# Create number of distict events for each category in categorical numeric variables
for variable_name, test in categorical_numeric.iterrows():
    #Frequency distribution
    distribution = data[variable_name].value_counts(dropna= False).to_dict()
    
    if np.nan in list(distribution.keys()):
        #Count NaN elements and add to distribution
        data_information['Nan_count'][variable_name] = distribution[np.nan]
    else:
        data_information['Nan_count'][variable_name] = 0
    
    # Setting with copy warning. Appears to have desired effect. 
    data_information['Category Distribution'][variable_name] = (list(distribution.keys()), list(distribution.values()))



/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  del sys.path[0]
/opt/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:205: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_with_indexer(indexer, value)
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/opt/anaconda3/lib/python3.7/site-pac

<b>Noncategorical Analysis</b>

For each noncategorical variable, summary statistics were created and stored. 

In [4]:
# Create summary statistics for numerics
for variable_name, notused in numeric.iterrows():
    #Store summary statistics
    statistics = data[variable_name].describe().to_dict()
    nan_count = data[variable_name].isna().sum()
    data_information['Summary Statistics'][variable_name] = (list(statistics.keys()), list(statistics.values()))
    data_information.loc[variable_name, 'Nan_count'] = data[variable_name].isna().sum()


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [5]:
# Determine the percentage of missing data in each variable 
data_information['Nan_percent'] = data_information['Nan_count']/len(data)

# View the dataframe created
data_information

Type  \
Column Name                          
MSSubClass     Categorical Numeric   
MSZoning        Categorical String   
LotFrontage                Numeric   
LotArea                    Numeric   
Street          Categorical String   
...                            ...   
MoSold         Categorical Numeric   
YrSold                     Numeric   
SaleType        Categorical String   
SaleCondition   Categorical String   
SalePrice                  Numeric   

                                           Category Distribution  Nan_count  \
Column Name                                                                   
MSSubClass     ([20, 60, 50, 120, 30, 160, 70, 80, 90, 190, 8...        0.0   
MSZoning       ([RL, RM, FV, RH, C (all)], [1151, 218, 65, 16...        0.0   
LotFrontage                                                  NaN      259.0   
LotArea                                                      NaN        0.0   
Street                                 ([Pave, Grvl], [1454, 6])        0.0   
...                                                          ...        ...   
MoSold         ([6, 7, 5, 4, 8, 3, 10, 11, 9, 12, 1, 2], [253...        0.0   
YrSold                                                       NaN        0.0   
SaleType       ([WD, New, COD, ConLD, ConLI, ConLw, CWD, Oth,...        0.0   
SaleCondition  ([Normal, Partial, Abnorml, Family, Alloca, Ad...        0.0   
SalePrice                                                    NaN        0.0   

               Nan_percent                                 Summary Statistics  
Column Name                                                                    
MSSubClass        0.000000                                                NaN  
MSZoning          0.000000                                                NaN  
LotFrontage       0.177397  ([count, mean, std, min, 25%, 50%, 75%, max], ...  
LotArea           0.000000  ([count, mean, std, min, 25%, 50%, 75%, max], ...  
Street            0.000000                                                NaN  
...                    ...                                                ...  
MoSold            0.000000                                                NaN  
YrSold            0.000000  ([count, mean, std, min, 25%, 50%, 75%, max], ...  
SaleType          0.000000                                                NaN  
SaleCondition     0.000000                                                NaN  
SalePrice         0.000000  ([count, mean, std, min, 25%, 50%, 75%, max], ...  

[80 rows x 5 columns]

In an effort to find missing values, we examine each variable that has occurances of NaN. 

In [6]:
data_information[data_information['Nan_count'] > 0]

Type  \
Column Name                        
LotFrontage              Numeric   
Alley         Categorical String   
MasVnrType    Categorical String   
MasVnrArea               Numeric   
BsmtQual      Categorical String   
BsmtCond      Categorical String   
BsmtExposure  Categorical String   
BsmtFinType1  Categorical String   
BsmtFinType2  Categorical String   
Electrical    Categorical String   
FireplaceQu   Categorical String   
GarageType    Categorical String   
GarageYrBlt              Numeric   
GarageFinish  Categorical String   
GarageQual    Categorical String   
GarageCond    Categorical String   
PoolQC        Categorical String   
Fence         Categorical String   
MiscFeature   Categorical String   

                                          Category Distribution  Nan_count  \
Column Name                                                                  
LotFrontage                                                 NaN      259.0   
Alley                       ([nan, Grvl, Pave], [1369, 50, 41])     1369.0   
MasVnrType    ([None, BrkFace, Stone, BrkCmn, nan], [864, 44...        8.0   
MasVnrArea                                                  NaN        8.0   
BsmtQual       ([TA, Gd, Ex, nan, Fa], [649, 618, 121, 37, 35])       37.0   
BsmtCond         ([TA, Gd, Fa, nan, Po], [1311, 65, 45, 37, 2])       37.0   
BsmtExposure  ([No, Av, Gd, Mn, nan], [953, 221, 134, 114, 38])       38.0   
BsmtFinType1  ([Unf, GLQ, ALQ, BLQ, Rec, LwQ, nan], [430, 41...       37.0   
BsmtFinType2  ([Unf, Rec, LwQ, nan, BLQ, ALQ, GLQ], [1256, 5...       38.0   
Electrical    ([SBrkr, FuseA, FuseF, FuseP, Mix, nan], [1334...        1.0   
FireplaceQu   ([nan, Gd, TA, Fa, Ex, Po], [690, 380, 313, 33...      690.0   
GarageType    ([Attchd, Detchd, BuiltIn, nan, Basment, CarPo...       81.0   
GarageYrBlt                                                 NaN       81.0   
GarageFinish        ([Unf, RFn, Fin, nan], [605, 422, 352, 81])       81.0   
GarageQual    ([TA, nan, Fa, Gd, Ex, Po], [1311, 81, 48, 14,...       81.0   
GarageCond    ([TA, nan, Fa, Gd, Po, Ex], [1326, 81, 35, 9, ...       81.0   
PoolQC                     ([nan, Gd, Ex, Fa], [1453, 3, 2, 2])     1453.0   
Fence         ([nan, MnPrv, GdPrv, GdWo, MnWw], [1179, 157, ...     1179.0   
MiscFeature   ([nan, Shed, Gar2, Othr, TenC], [1406, 49, 2, ...     1406.0   

              Nan_percent                                 Summary Statistics  
Column Name                                                                   
LotFrontage      0.177397  ([count, mean, std, min, 25%, 50%, 75%, max], ...  
Alley            0.937671                                                NaN  
MasVnrType       0.005479                                                NaN  
MasVnrArea       0.005479  ([count, mean, std, min, 25%, 50%, 75%, max], ...  
BsmtQual         0.025342                                                NaN  
BsmtCond         0.025342                                                NaN  
BsmtExposure     0.026027                                                NaN  
BsmtFinType1     0.025342                                                NaN  
BsmtFinType2     0.026027                                                NaN  
Electrical       0.000685                                                NaN  
FireplaceQu      0.472603                                                NaN  
GarageType       0.055479                                                NaN  
GarageYrBlt      0.055479  ([count, mean, std, min, 25%, 50%, 75%, max], ...  
GarageFinish     0.055479                                                NaN  
GarageQual       0.055479                                                NaN  
GarageCond       0.055479                                                NaN  
PoolQC           0.995205                                                NaN  
Fence            0.807534                                                NaN  
MiscFeature      0.963014                           

Each of these variables was looked at further to determine what to do with the instances of NaN. 

<b>Categorical NaN Investigations and decisions</b>

Alley has 1369 instances of NaN that are addressed in the data documentation as 'No Alley Access'. These instances will be categorized reflect that. 

BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1 all have NaN values. These instances are meant to mean no basement. The code below will update NaN values to reflect a category of 'No Basement'. Similarly, BsmtFinType2 NaN occourances reflect no feature, and will be categorized as such. 

GarageType, GarageFinish, GarageCond, and GarageQual all have 81 instances of NaN values. These instances are meant to mean no garage. All three variables agree on the no garage count. Category title will be changed from NaN to 'No Garage'

MasVnrType contains 8 instances of NaN. These instances will be left as NaN as they are truly unknow. This variable already has a category for intances of 'None', and therefore we do not want to confuse those with 'Unknown'.

Electrical is a categorical describing the type of electrical system. There is one instance of NaN, taken to mean unknown. This value will be left as NaN. 

FireplaceQu is a categorical variable that describes the quality of the fireplace. NaN instances taken to mean no fireplace. Category title will be updated from NaN to 'No Fireplace'.

PoolQC is a categorical describing the quality of the pool. Greater than 99% of the data is NaN, taken to mean no pool. Category will be updated from NaN to 'No Pool'.

Fence is a categorical describing the quality of the fence. 80% of the data is NaN taken to mean no fence. Category title will be updated from NaN to 'No Fence'.

MiscFeature is a categorical used to capture miscelaneous features of the property that didn't occur in one of the other variables. The 1406 instances of NaN mean no feature measured, and NaN is an acceptable way to leave that. 

In [7]:
## Updating the raw data file

#Alley
data.loc[data['Alley'].isna(), 'Alley'] = 'No Alley'

# BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinType2
data.loc[data['BsmtQual'].isna(), 'BsmtQual'] = 'No BsmtQual'
data.loc[data['BsmtCond'].isna(), 'BsmtCond'] = 'No Basement'
data.loc[data['BsmtExposure'].isna(), 'BsmtExposure'] = 'No Basement'
data.loc[data['BsmtFinType1'].isna(), 'BsmtFinType1'] = 'No Basement'
data.loc[data['BsmtFinType2'].isna(), 'BsmtFinType2'] = 'No Basement'

# GarageType, GarageFinish, GarageQual, and GarageCond
data.loc[data['GarageType'].isna(), 'GarageType'] = 'No Garage'
data.loc[data['GarageFinish'].isna(), 'GarageFinish'] = 'No Garage'
data.loc[data['GarageQual'].isna(), 'GarageQual'] = 'No Garage'
data.loc[data['GarageCond'].isna(), 'GarageCond'] = 'No Garage'

# FireplaceQu, PoolQC, Fence
data.loc[data['FireplaceQu'].isna(), 'FireplaceQu'] = 'No Fireplace'
data.loc[data['PoolQC'].isna(), 'PoolQC'] = 'No Pool'
data.loc[data['Fence'].isna(), 'Fence'] = 'No Fence'

## Updating the data_information dataframe
for variable_name, test in categorical_string.iterrows():
    distribution = data[variable_name].value_counts(dropna= False).to_dict()
    
    if np.nan in list(distribution.keys()):
        #Count NaN elements and add to distribution
        data_information.loc[variable_name, 'Nan_count'] = distribution[np.nan]
    else:
        #data_information['Nan_count'][variable_name] = 0
        data_information.loc[variable_name, 'Nan_count'] = 0
    
    # This line of code works as intended but throws SetWithCopy warning.
    data_information['Category Distribution'][variable_name] = (list(distribution.keys()), list(distribution.values()))


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:36: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


<b>Noncategorical NaN investigation</b>

There were three noncategorical variables that had missing values. 

Lot Frontage had 259 missing values. These missing values were imputed using a mean. Median was also considered, but both mean and median were very close to eachother and mean was used. 

MasVnrArea had 6 instances of NaN. These instances were left as NaN as they correspond with the 8 instances of MasVnrType that are also NaN. That categorical already contains a 'None' category, with MasVnrArea values of zero. This is notably different. These datapoints will be considered for removal.  

GarageYrBlt also has 81 instances of NaN. These will be left as NaN as they coorespond to the 81 data points that do not have garages. A value of 0 would imply the existance of a garage with a build year over 2000 years ago. NaN is more approprate for no garage.  


In [8]:
# Checking summary statistics
print(data['LotFrontage'].describe())

count    1201.000000
mean       70.049958
std        24.284752
min        21.000000
25%        59.000000
50%        69.000000
75%        80.000000
max       313.000000
Name: LotFrontage, dtype: float64


In [9]:
# Replacing NaN values with mean
data.loc[data['LotFrontage'].isna(), 'LotFrontage'] = data['LotFrontage'].mean()

# Resetting NaN Count
data_information.loc['LotFrontage', 'Nan_count'] = data['LotFrontage'].isna().sum()

In [10]:
# Save off data and data_information
data.to_csv('Housing_Wrangled.csv')
data_information.to_csv('data_information.csv')